content/machine-learning/regreg.Rmarkdown

```diff
@@ -0,0 +1,129 @@
+
+```{r, message=FALSE}
+require(dplyr)
+require(quanteda)
+require(quanteda.textmodels)
+require(glmnet)
+require(caret)
+```
+
+`data_corpus_moviereviews` from the **quanteda.textmodels** package contains 2000 movie reviews classified either as "positive" or "negative".
+
+```{r}
+corp_movies <- data_corpus_moviereviews
+summary(corp_movies, 5)
+```
+
+The variable "Sentiment" indicates whether a movie review was classified as positive or negative. In this example we use 1500 reviews as the training set and build a regularized regression classifier based on this subset. In a second step, we predict the sentiment for the remaining reviews (our test set).
+
+Since the first 1000 reviews are negative and the remaining reviews are classified as positive, we need to draw a random sample of the documents.
+
+```{r}
+# generate 1500 numbers without replacement
+set.seed(300)
+id_train <- sample(1:2000, 1500, replace = FALSE)
+head(id_train, 10)
+
+# create docvar with ID
+corp_movies$id_numeric <- 1:ndoc(corp_movies)
+
+# get training set
+dfmat_training <- corpus_subset(corp_movies, id_numeric %in% id_train) %>%
+    dfm(remove = stopwords("en"), stem = TRUE)
+
+# get test set (documents not in id_train)
+dfmat_test <- corpus_subset(corp_movies, !(id_numeric %in% id_train)) %>%
+    dfm(remove = stopwords("en"), stem = TRUE)
+```
+
+Next we choose $\lambda$ using `cv.glmnet` from the **glmnet** package. `cv.glmnet` requires an input matrix X and a response vector Y. For the input matrix, we use the training set converted to a sparse matrix. For the response vector, we use a dichotomous indicator of review sentiment in the training set with negative reviews coded as 1 and positive reviews as 0. 
+
+`cv.glmnet` uses cross-validation to select the value of $\lambda$ yielding the smallest classification error. Setting `alpha=1` selects the lasso estimator. Setting `nfold=5` sets cross-validation across 5 'folds', i.e. partitions of the training set.
+
+```{r}
+trainingmat <- as(dfmat_training, "dgCMatrix")
+trainingyvec <- ifelse(dfmat_training$sentiment == "neg", 1, 0)
+
+lasso <- cv.glmnet(x = trainingmat,
+                   y = trainingyvec,
+                   alpha = 1,
+                   nfold = 5,
+                   family = "binomial")
+```
+
+As an initial evaluation of the model, let's take a look at the most predictive features. We begin by obtaining the best value of the penalty parameter $\lambda$:
+
+```{r}
+best.lambda <- which(lasso$lambda == lasso$lambda.min)
+beta <- lasso$glmnet.fit$beta[, best.lambda]
+```
+
+We can now look at the most predictive features for this value of $\lambda$:
+
+```{r}
+mostpreddf <- data.frame(
```

Comment:
Data.frame should be `dat_xxx`.

Reply:
Just to make sure: you mean the object name should be of the form `dat_xxx`, e.g. `dat_mostpred`?

Reply:
Yes, `dat_mostpred` follows the naming convention. `dat_pred` would be enough.